### PR TITLE
kernel: grant: do not pass T::default()

### PR DESCRIPTION
**Pull request by bradjc**
This changes how grant allocation is structured in grant.rs so that the
allocation function does not take a copy of the data to be written when
creating a new grant. This data can be large if the grant region is
large, and it can cause a stack overflow when the allocate function is
called.

The change is pretty simple, but I'm not sure if there is some subtle reason to not do it this way.

Fixes the IPC stack overflow issue found in #1933. Replaces #1976.

### Testing Strategy

Running the hail app on hail.

### TODO or Help Wanted

n/a

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: kernel/src/grant.rs [hunk 1]
```diff
@@ -102,7 +102,7 @@ impl Allocator {
 
                         // We use `ptr::write` to avoid `Drop`ping the uninitialized memory in
                         // case `T` implements the `Drop` trait.
-                        write(ptr.as_ptr(), data);
+                        write(ptr.as_ptr(), T::default());
```
Comment (brghena) [inline kernel/src/grant.rs:105]: I'm pretty uncertain why this change would actually solve the "grants temporarily take up a ton of stack space" problem. `T::default()` is still going to return a T, which will be its full size and on the stack. Is passing it into the `alloc_unowned()` function taking double the size of T on the stack or something?

Comment (brghena) [reply kernel/src/grant.rs:105]: Per a slack discussion with bradjc, this change greatly increases the possible size of IPC grants before a kernel stack overflow occurs (previously 364 bytes increases to 2364 bytes). I'm guessing some compiler optimizations are making the magic happen here.

Comment (alevy) [reply kernel/src/grant.rs:105]: Right, presumably the compiler can better guess to just allocate the space for the default value directly at `ptr.as_ptr()` as opposed to on the stack first.

**Pull request merged**